homeassistant/components/swiss_public_transport/__init__.py

```diff
@@ -1 +1,64 @@
 """The swiss_public_transport component."""
+import logging
+
+from opendata_transport import OpendataTransport
+from opendata_transport.exceptions import OpendataTransportError
+import voluptuous as vol
+
+from homeassistant import config_entries, core
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import Platform
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+import homeassistant.helpers.config_validation as cv
+
+from .const import CONF_DESTINATION, CONF_START, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_START): cv.string,
+        vol.Required(CONF_DESTINATION): cv.string,
+    }
+)
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+
+async def async_setup_entry(
+    hass: core.HomeAssistant, entry: config_entries.ConfigEntry
+) -> bool:
+    """Set up platform from a ConfigEntry."""
+    config = entry.data
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = config
+
+    start = config.get(CONF_START)
+    destination = config.get(CONF_DESTINATION)
+
+    session = async_get_clientsession(hass)
+    opendata = OpendataTransport(start, destination, session)
+
+    try:
+        await opendata.async_get_data()
+    except OpendataTransportError:
+        _LOGGER.error(
+            "Check at http://transport.opendata.ch/examples/stationboard.html "
```

Comment:
I think there are 2 things we should try and catch. Because we already check if the stations exist in the config flow, we know that _in theory_ the stations exist and that this can work. But there are 2 possibilities where we can't get data.

1. The service is down. In this case we want to raise a ConfigEntryNotReady exception. This makes it retry a bit later again.
2. The stations don't exist anymore. (I'm just doing a guess this is something that can happen). In this case, our config entry just doesn't work. In the future, if you feel really fancy you could work on raising a repair issue to notify the user that they should repair their config entry. For now, please log that the call failed and raise ConfigEntryError instead.

This is the ideal solution, but this also requires the library to raise different issues on the different causes. Is there any way you can maybe tinker around with this and check if we can catch specifics?

Reply:
Yes, I see the point. There are two errors thrown by the library, one for connection problems: `OpendataTransportConnectionError`; and one, when he cannot interpret the response `OpendataTransportError`.

It can be found here:
https://github.com/home-assistant-ecosystem/python-opendata-transport/blob/2077fe0bfe4e4e71554d6d364c2b4a30c9729863/opendata_transport/__init__.py#L290

But I also managed to get a `KeyError` out of the library, which the library did not catch, and I would therefore propose to also add a catch all at the end for _other_ errors.

Reply:
Beware, we may not do `except Exception:`, that's only allowed in the config flow.

Reply:
Ok, so I will only take care of the Errors the library provides.

Reply:
But, you can catch a KeyError (and I hope you will update the lib to catch this error properly in the lib and patch it up here)

Reply:
I realized, I only got the KeyError because I used the `AiohttpClientMocker` in the unit test and provided invalid data. I would rather not implement the KeyError and patch the library to also catch KeyErrors.